src/topic.rs

```diff
@@ -30,51 +28,50 @@ use crate::interface::mapping::path::{MappingError, MappingPath};
 pub enum TopicError {
     #[error("topic is empty")]
     Empty,
+    #[error(
+        "the topic should start with <realm>/<device_id> equal to {client_id}, received: {topic}"
+    )]
+    UnkownClientId { client_id: String, topic: String },
     #[error(
         "the topic should be in the form <realm>/<device_id>/<interface>/<path>, received: {0}"
     )]
     Malformed(String),
-
-    #[error("couldn't parse mapping for '{topic}'")]
-    Maapping {
-        #[source]
-        err: MappingError,
-        topic: String,
-    },
 }
 
 impl TopicError {
     pub fn topic(&self) -> &str {
         match self {
             TopicError::Empty => "",
+            TopicError::UnkownClientId { topic, .. } => topic,
             TopicError::Malformed(topic) => topic,
-            TopicError::Maapping { topic, .. } => topic,
         }
     }
 }
 
-pub(crate) fn parse_topic(topic: &str) -> Result<(&str, &str, &str, MappingPath), TopicError> {
+#[derive(Debug)]
+pub(crate) struct ParsedTopic<'a> {
+    pub(crate) interface: &'a str,
+    pub(crate) path: &'a str,
+}
+
+pub(crate) fn parse_topic<'a>(
```

Comment:
Why isn't this a constructor for the PardetTopic struct. It could be named `try_parse()`

Reply:
You're right, I'll refactor the code like you suggested, thanks.